### ISSUE TITLE
Windows Junctions inside the Recycle Bin are being followed through under certain circumstances.

### ISSUE THREAD

**Issue by SenpaiSharp** [bleachbit/bleachbit#668]
(Using Windows' terminology, as mklink describes)

It doesn't seem to affect symbolic links but if you put a **junction** inside of a folder and then put that folder into the recycle bin, Bleachbit will follow through into the folder and _into_ the junction and mark everything inside of the junction's target for deletion.

**Comment by az0** [comment]
@SenpaiSharp  Thanks for the report. I tried to reproduce it, but I couldn't, so would you please send me some more info?

1. Start BleachBit
1. Click the application icon to open up the drop-down menu
1. Click diagnostics
1. Copy either all the lines. (If you wish to remove the lines with pathnames, that's fine.)
1. Paste that information here into a GitHub comment

**Comment by Tobias-B-Besemer** [comment]
@SenpaiSharp, for which user(s) do you want to delete the Recycle Bin?
For just your current, or for all? (Needs admin rights...)
On drive C:, or on other drives, too?
Windows (System) cleaning is ATM done from within the core app. I made an attemp to further cleN Windows with a CleanerML file...
The CleanerML file deletes the Recycle Bin in a other way, but with the cleaner following for all users and only on drive C:.
But you can download and test it!
Here you can find the CleanerML file:
https://github.com/az0/cleanerml/pull/170

I made a second cleaner just for the Recycle Bin.
He deletes the bin of all users but on all drives!
You can find it here:
https://github.com/az0/cleanerml/blob/master/pending/windows_recycle_bins_more.xml

A feedback to one or both cleaners would be really welcome!

@az0, if one of my cleaners solves the issue, time to let the first parts of Windows be cleaned with CleanerML ???

**Comment by az0** [comment]
@Tobias-B-Besemer  
I would not recommend that CleanerML cleaner for cleaning a recycle bin that contains a symlink or a junction because it won't check for them. BleachBit checks for these special cases only when using the built-in recycle bin cleaning method found under the System category.

**Comment by Tobias-B-Besemer** [comment]
@az0, the CleanerML cleaner deletes with an action command="delete", search="global" and as far as I understand it, this action don't/can't follow links...

**Comment by SenpaiSharp** [comment]
> 
> 
> @SenpaiSharp Thanks for the report. I tried to reproduce it, but I couldn't, so would you please send me some more info?
> 
>     1. Start BleachBit
> 
>     2. Click the application icon to open up the drop-down menu
> 
>     3. Click diagnostics
> 
>     4. Copy either all the lines. (If you wish to remove the lines with pathnames, that's fine.)
> 
>     5. Paste that information here into a GitHub comment

**BleachBit Diagnostics** 

```
BleachBit version 3.0
Git revision c32a0365
GTK version 3.18.9
SQLite version 3.28.0
FSE = mbcs
local_cleaners_dir = C:\Portable Programs\BleachBit-Portable\cleaners
locale_dir = C:\Portable Programs\BleachBit-Portable\share\locale\
options_dir = C:\Portable Programs\BleachBit-Portable
personal_cleaners_dir = C:\Portable Programs\BleachBit-Portable\cleaners
system_cleaners_dir = C:\Portable Programs\BleachBit-Portable\share\cleaners\
locale.getdefaultlocale = ('en_US', 'cp1252')
os.getenv('cd') = C:\Portable Programs\BleachBit-Portable
os.getenv('ProgramFiles') = C:\Program Files (x86)
os.getenv('ProgramW6432') = C:\Program Files
platform.win32_ver[1]() = 6.3.9600
platform.platform = Windows-8.1-6.3.9600
platform.version = 6.3.9600
sys.argv = ['C:\\Portable Programs\\BleachBit-Portable\\bleachbit.exe', '--gui', '--no-uac']
sys.executable = C:\Portable Programs\BleachBit-Portable\bleachbit.exe
sys.version = 2.7.16 (v2.7.16:413a49145e, Mar  4 2019, 01:30:55) [MSC v.1500 32 bit (Intel)]
win32com.shell.shell.IsUserAnAdmin() = True
__file__ = C:\Portable Programs\BleachBit-Portable\library.zip\bleachbit\Diagnostic.pyo
```

**Command Line Test**

```
mkdir target

mkdir container

mkdir target\DoesBBSeeThis

mklink /J container\link target
```

![containertorecycle](https://user-images.githubusercontent.com/36146858/68212017-b6521f00-ffa6-11e9-8b77-80b077ef11a7.png)

**BleachBit Preview**

```
Delete 0 A:\$RECYCLE.BIN\S-1-5-21-435774263-881465020-3127623276-1001\$RO95MXP\link\DoesBBSeeThis
Delete 0 A:\$RECYCLE.BIN\S-1-5-21-435774263-881465020-3127623276-1001\$RO95MXP\link
Delete 0 A:\$RECYCLE.BIN\S-1-5-21-435774263-881465020-3127623276-1001\$RO95MXP
Empty the recycle bin ?B 

Disk space to be recovered: 0
Files to be deleted: 3
Special operations: 1
```

**Comment by az0** [comment]
@SenpaiSharp 
Your comment helped me make progress

**Comment by az0** [comment]
The problem is that even Python 3.7's `os.walk()`, which BleachBit uses, follows junctions (Python [issue 23407](https://bugs.python.org/issue23407)). 

I'll keep working on it.

**Comment by Tobias-B-Besemer** [comment]
> 
> 
> @az0, the CleanerML cleaner deletes with an action command="delete", search="global" and as far as I understand it, this action don't/can't follow links...

Do my solution follow junctions, too ???

**Comment by az0** [comment]
@Tobias-B-Besemer 
Yes

**Comment by az0** [comment]
I published this fix in [BleachBit 3.1.0 beta](https://www.bleachbit.org/news/bleachbit-31-beta)
